### ISSUE TITLE
[Android]Rendering takes longer when drawing lines with huge data points

### ISSUE THREAD

**Issue by DeviArunaMurugan** [flutter/flutter#95625]
I have drawn line path using the canvas.drawLine() with huge amount of data points(30K points). I have taken time for calculating the points and rendering the time by using the Stopwatch. But once after the points are calculated and the canvas.drawLine are called, the lines are rendered with the delay in view. How to reduce those time delay ?

```dart
import 'dart:math';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: CustomPaint(
          painter: LinePainter(),
          child: Container(),
        ),
      ),
    );
  }
}

class LinePainter extends CustomPainter {
  late List<LineData> dataPoints;
  LinePainter() {
    dataPoints = <LineData>[];
    Random random = Random();
    for (int i = 0; i < 30000; i++) {
      final int y = random.nextInt(100);
      dataPoints.add(LineData(i, y));
    }
  }
  @override
  void paint(Canvas canvas, Size size) {
    final Stopwatch watch = Stopwatch();
    watch.start();
    LineData currentPoint, _nextPoint;
    Offset? startPoint, endPoint;
    for (int pointIndex = 0; pointIndex < dataPoints.length; pointIndex++) {
      currentPoint = dataPoints[pointIndex];
      startPoint = pointToOffset(currentPoint, size);

      if (pointIndex + 1 < dataPoints.length) {
        _nextPoint = dataPoints[pointIndex + 1];
        endPoint = pointToOffset(_nextPoint, size);
      }

      if (endPoint != null) {
        canvas.drawLine(startPoint, endPoint, getStrokePaint());
        endPoint = startPoint = null;
      }
    }

    watch.stop();
    print("Timer: ");
    print(watch.elapsedMilliseconds);
  }

  Paint getStrokePaint() {
    final Paint strokePaint = Paint();
    strokePaint.strokeWidth = 2;
    strokePaint.style = PaintingStyle.fill;
    strokePaint.strokeCap = StrokeCap.round;
    strokePaint.color = Colors.black;
    return strokePaint;
  }

  Offset pointToOffset(LineData data, Size size) {
    num x = valueToXCoefficient(data.x);
    num y = valueToYCoefficient(data.y);
    final num xLength = size.width;
    final num yLength = size.height;
    final double locationX = (x * xLength).toDouble();
    final double locationY = ((1 - y) * yLength).abs().toDouble();
    return Offset(locationX, locationY);
  }

  num valueToXCoefficient(num value) {
    num result = 0;
    result = value / (dataPoints.length / 2);

    return result;
  }

  num valueToYCoefficient(num value) {
    num result = 0;
    result = value / 50;
    return result;
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }
}

class LineData {
  LineData(this.x, this.y);
  final int x;
  final int y;
}
```

Regards

**Comment by maheshmnj** [comment]
Hi @DeviArunaMurugan, Thanks for filing the issue. I ran your sample code and it roughly took 39ms on Average

```
flutter: Timer:
flutter: 35
flutter: Timer:
flutter: 37
Reloaded 1 of 772 libraries in 752ms.
flutter: Timer:
flutter: 36
Reloaded 1 of 772 libraries in 537ms.
flutter: Timer:
flutter: 41
Reloaded 1 of 772 libraries in 528ms.
```
 
I believe this is working as intended to further optimize this I would suggest you try asking on support channels like StackOverflow. Closing as this is neither a bug nor a feature request.

**Comment by DeviArunaMurugan** [comment]
@maheshmnj , May I know, have you referred the time average, which I have used in the code? if so, I have taken average for the above code is 289.2(375, 272, 260, 255, 284). I have mentioned regarding the loading time delay after the execution of the for loop(point and rendering related code).

**Comment by DeviArunaM** [comment]
Hi @maheshmnj : We have checked more than four time in various high configuration machine. In that case , we are getting average timing as 290ms. Could you share in which case, your average rendering timing as 39ms?

Regards,

**Comment by maheshmnj** [comment]
@DeviArunaMurugan as mentioned above 39ms is the average rendering time I am seeing for the above logs shared (35,37,36,41)ms. Can you please share your logs for the above code sample and the `flutter doctor -v` to investigate further? Also for which platform are you compiling?
 
Thanks

**Comment by DeviArunaMurugan** [comment]
@maheshmnj : Please find the flutter doctor logs,

PS D:\sample\line_performance> flutter doctor -v
[√] Flutter (Channel stable, 2.8.0, on Microsoft Windows [Version 10.0.19043.1415], locale en-US)
    • Flutter version 2.8.0 at C:\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision cf44000065 (2 weeks ago), 2021-12-08 14:06:50 -0800
    • Engine revision 40a99c5951
    • Dart version 2.15.0

[!] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at C:\Users\TempKaruna\AppData\Local\Android\sdk
    X cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    X Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Professional 2022 17.0.2)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Professional
    • Visual Studio Professional 2022 version 17.0.31919.166
    • Windows 10 SDK version 10.0.22000.0

[√] Android Studio (version 2020.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[√] VS Code (version 1.63.2)
    • VS Code at C:\Users\TempKaruna\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.29.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19043.1415]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 96.0.4664.110
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 96.0.1054.53
    ! Device ZY223G4JQF is offline.


I am compiling in android and please find the logs

```
I/flutter (31642): Timer:
I/flutter (31642): 264
I/flutter (31838): Timer:
I/flutter (31838): 338
I/flutter (32017): Timer:
I/flutter (32017): 260
I/flutter (32297): Timer:
I/flutter (32297): 293
I/flutter (32441): Timer:
I/flutter (32441): 277
```

Average is 286.4

Regards,

**Comment by maheshmnj** [comment]
@DeviArunaMurugan I verified this again on Xiaomi k20 pro (Android) in release mode These are the rendering times I get on hot restarting the app. These logs are from debug mode.

```
Restarted application in 2,556ms.
I/flutter (29099): Timer:
I/flutter (29099): 66
I/flutter (29099): Timer:
I/flutter (29099): 62
Reloaded 1 of 567 libraries in 923ms.
I/flutter (29099): Timer:
I/flutter (29099): 53
Reloaded 1 of 567 libraries in 708ms.
I/flutter (29099): Timer:
I/flutter (29099): 61
Reloaded 1 of 567 libraries in 643ms.
I/flutter (29099): Timer:
I/flutter (29099): 69
Reloaded 1 of 567 libraries in 561ms.
I/flutter (29099): Timer:
I/flutter (29099): 57
Reloaded 1 of 567 libraries in 539ms.
```

In what mode are you running the app (debug/release/profile)? 
Also are you verifying this on a simulator or a physical device? if the latter please name the device

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.8.1, on macOS 12.1 21C52 darwin-arm, locale en-GB)
    • Flutter version 2.8.1 at /Users/mahesh/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (7 days ago), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 60.1.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.61.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (3 available)
    • Redmi K20 Pro (mobile) • 192.168.1.7:5555 • android-arm64  • Android 11 (API 30)
    • macOS (desktop)        • macos            • darwin-arm64   • macOS 12.1 21C52 darwin-arm
    • Chrome (web)           • chrome           • web-javascript • Google Chrome 96.0.4664.110

• No issues found!
```
</details>

Additionally, try upgrading to stable 2.8.1 and see if there is any difference.

**Comment by DeviArunaMurugan** [comment]
@maheshmnj : I have upgraded to flutter 2.8.1 and taken the below metrics with hot reloaded the app in debug mode.

```
I/flutter ( 7395): Timer:
I/flutter ( 7395): 270
I/flutter ( 7395): Timer:
I/flutter ( 7395): 255
Reloaded 1 of 567 libraries in 2,408ms.
I/flutter ( 7395): Timer:
I/flutter ( 7395): 250
Reloaded 1 of 567 libraries in 1,939ms.
I/flutter ( 7395): Timer:
I/flutter ( 7395): 239
Reloaded 1 of 567 libraries in 1,992ms.
I/flutter ( 7395): Timer:
I/flutter ( 7395): 243
Reloaded 1 of 567 libraries in 2,225ms.
```

please find the log after flutter upgrade,

```
[√] Flutter (Channel stable, 2.8.1, on Microsoft Windows [Version 10.0.19043.1415], locale en-US)
    • Flutter version 2.8.1 at C:\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (7 days ago), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1

[!] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at C:\Users\TempKaruna\AppData\Local\Android\sdk
    X cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.   
    X Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Professional 2022 17.0.2)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Professional
    • Visual Studio Professional 2022 version 17.0.31919.166
    • Windows 10 SDK version 10.0.22000.0

[√] Android Studio (version 2020.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[√] VS Code (version 1.63.2)
    • VS Code at C:\Users\TempKaruna\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.29.0

[√] Connected device (4 available)
    • Moto G 4 (mobile) • ZY223G4JQF • android-arm    • Android 7.0 (API 24)
    • Windows (desktop) • windows    • windows-x64    • Microsoft Windows [Version 10.0.19043.1415]
    • Chrome (web)      • chrome     • web-javascript • Google Chrome 96.0.4664.110
    • Edge (web)        • edge       • web-javascript • Microsoft Edge 96.0.1054.53

```
I have taken the above metrics using moto G4.

Regards,

**Comment by maheshmnj** [comment]
reopening the issue based on the above comment

**Comment by maheshmnj** [comment]
@DeviArunaMurugan, I am unable to reproduce the issue, Can you please share the timeline trace (JSON file) by following this guide https://medium.com/flutter/profiling-flutter-applications-using-the-timeline-a1a434964af3#a499, Also is it possible for you to verify on another Android device and share the logs 

<details>
<summary>master channel logs</summary>

```
mahesh@Maheshs-MacBook-Air-M1 sample % flutterm run
Could not load custom device from config index 0: Expected enabled to be a boolean.
Launching lib/main.dart on Redmi K20 Pro in debug mode...
Warning: Mapping new ns http://schemas.android.com/repository/android/common/02 to old ns http://schemas.android.com/repository/android/common/01
Warning: Mapping new ns http://schemas.android.com/repository/android/generic/02 to old ns http://schemas.android.com/repository/android/generic/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/addon2/02 to old ns http://schemas.android.com/sdk/android/repo/addon2/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/repository2/02 to old ns http://schemas.android.com/sdk/android/repo/repository2/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/sys-img2/02 to old ns http://schemas.android.com/sdk/android/repo/sys-img2/01
Warning: Mapping new ns http://schemas.android.com/repository/android/common/02 to old ns http://schemas.android.com/repository/android/common/01
Warning: Mapping new ns http://schemas.android.com/repository/android/generic/02 to old ns http://schemas.android.com/repository/android/generic/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/addon2/02 to old ns http://schemas.android.com/sdk/android/repo/addon2/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/repository2/02 to old ns http://schemas.android.com/sdk/android/repo/repository2/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/sys-img2/02 to old ns http://schemas.android.com/sdk/android/repo/sys-img2/01
Running Gradle task 'assembleDebug'...                             21.0s
✓  Built build/app/outputs/flutter-apk/app-debug.apk.
Installing build/app/outputs/flutter-apk/app.apk...                 7.8s
I/flutter (31296): Timer: 
I/flutter (31296): 45
Syncing files to device Redmi K20 Pro...                           202ms
I/.example.sampl(31296): ProcessProfilingInfo new_methods=1020 is saved saved_to_disk=1 resolve_classes_delay=8000

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h List all available interactive commands.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).

💪 Running with sound null safety 💪

An Observatory debugger and profiler on Redmi K20 Pro is available at: http://127.0.0.1:50764/FoZTLku0NoQ=/
The Flutter DevTools debugger and profiler on Redmi K20 Pro is available at:
http://127.0.0.1:9102?uri=http://127.0.0.1:50764/FoZTLku0NoQ=/

I/flutter (31296): Timer: 
I/flutter (31296): 33
Performing hot reload...                                                
Reloaded 0 libraries in 739ms.

I/flutter (31296): Timer: 
I/flutter (31296): 30
Performing hot reload...                                                
Reloaded 0 libraries in 319ms.

I/flutter (31296): Timer: 
I/flutter (31296): 27
Performing hot reload...                                                
Reloaded 0 libraries in 184ms.

I/flutter (31296): Timer: 
I/flutter (31296): 29
Performing hot reload...                                                
Reloaded 0 libraries in 186ms.

I/flutter (31296): Timer: 
I/flutter (31296): 27
Performing hot reload...                                                
Reloaded 0 libraries in 233ms.

I/flutter (31296): Timer: 
I/flutter (31296): 29
Performing hot reload...                                                
Reloaded 0 libraries in 253ms.

I/flutter (31296): Timer: 
I/flutter (31296): 28
Performing hot reload...                                                
Reloaded 0 libraries in 237ms.

I/flutter (31296): Timer: 
I/flutter (31296): 29
Performing hot reload...                                                
Reloaded 0 libraries in 216ms.

I/flutter (31296): Timer: 
I/flutter (31296): 26
Performing hot reload...                                                
Reloaded 0 libraries in 327ms.
```
</details>
